File: translations/import.rb
require 'json'
require 'optparse'

unless File.exists? 'translations/projects.json'
    abort 'OH NO! projects.json cannot be found. Please run this script only from the root of the repository.'
end

options = {}
OptionParser.new do |opts|
  opts.banner = 'Usage: import.rb [options]'

  opts.on('-s', '--skipPull', 'Skips pulling from transifex') do |v|
    options[:skipPull] = v
  end

  opts.on('-pPROJECT', '--project=PROJECT', 'Import only a specific project') do |p|
    options[:project] = p
  end

  opts.on('-l', '--list', 'List projects that can be imported') do |l|
    options[:list] = l
  end

end.parse!

json = IO.read('translations/projects.json', encoding:'utf-8')
projects = JSON.parse json

if options[:list]
    puts projects.map { |project| project.fetch('name') }
    abort
end

# Gets everything that's translated from Transifex
unless options[:skipPull]
    command = 'tx pull -af'
    if options[:project]
        command << "r canvas-ios.en_#{options[:project]}*"
    end
    puts 'Importing everything from Transifex, this could take a while...'
    success = system(command)
    raise 'tx pull failed because reasons' unless success
end

projects = projects.select { |project| project.fetch('name') == options[:project] } if options[:project]
raise 'no projects to import' unless projects.count > 0
projects.each do |project|
    project_file = project.fetch('location')
    name = project.fetch('name')
    is_json = project.fetch('json')
    if is_json
        Dir.glob("#{project_file}/*.json") do |file|
            next if file.include? "temp.json"

            success = system("mv #{file} #{project_file}/temp.json && touch #{file}")
            raise "Failed to make a copy of #{file}" unless success

            temp = File.open("#{project_file}/temp.json")
            output = File.open(file, 'a')
            temp.each_line do |line|
                message = line[/^\s*\"message\": \"(.*)\"$/, 1].to_s

                if message.length > 0
                    message.gsub!('\"', '"')
                    message.gsub!('"', '\"')
                    line = "\t\"message\": \"#{message}\""
                end
                output.puts line
            end
            system("rm #{project_file}/temp.json")
        end
    else
        folder = "translations/canvas-ios.en_#{name}xliff"

        Dir.glob("#{folder}/*.xlf") do |file|
            puts "Importing #{file} into #{project_file}"
            success = system(%Q(xcodebuild -importLocalizations -localizationPath "#{file}" -project "#{project_file}"))
            raise 'xcodebuild -importLocalizations failed for some reason... :(' unless success
        end
    end
end

File: translations/import.js
#!/usr/bin/env node

const program = require('commander')
const { spawn } = require('child_process')
const { createReadStream, readFileSync, writeFileSync, readdir } = require('fs')
const mkdirp = require('mkdirp')
const path = require('path')
const S3 = require('aws-sdk/clients/s3')
const projects = require('./projects.json')

program
  .version(require('./package.json').version)
  .option('-s, --skipPull', 'Skip pulling from S3')
  .option('-p, --project [name]', 'Import only a specific project')
  .option('-l, --list', 'List projects that can be imported')

program.on('--help', () => {
  console.log(`
  Environment Variables:

    AWS_ACCESS_KEY_ID      AWS key, required to sync to instructure-translations S3 bucket
    AWS_SECRET_ACCESS_KEY  AWS secret, required to sync to instructure-translations S3 bucket
  \n`)
})
program.parse(process.argv)

if (program.list) {
  console.log(projects.map(({ name }) => name).join('\n'))
  process.exit(0)
}

if (
  !program.skipPull &&
  (!process.env.AWS_ACCESS_KEY_ID || !process.env.AWS_SECRET_ACCESS_KEY)
) {
  program.outputHelp()
  process.exit(1)
}

importTranslations().catch(err => {
  console.error(err)
  process.exit(2)
})

function run(cmd, args, opts) {
  return new Promise((resolve, reject) => {
    const command = spawn(cmd, args, opts)
    command.on('error', reject)
    command.on('exit', code => {
      if (code === 0) return resolve()
      console.error(command.stderr.toString())
      reject(`${cmd} failed with code ${code}.`)
    })
  })
}

async function importTranslations() {
  if (!program.skipPull) {
    const Bucket = 'instructure-translations'
    const s3 = new S3({ region: 'us-east-1' })
    const listObjects = await s3.listObjectsV2({ Bucket, Prefix: `canvas-ios/` }).promise()
    const keys = listObjects.Contents.map(({ Key }) => Key)
      .filter(key => !key.includes('/en/'))
      .filter(key => !program.project || key.includes(`/${program.project}.`))

    for (const key of keys) {
      const [ , locale, basename ] = key.split('/')
      const [ projectName, ext ] = basename.split('.')
      const filename = `imports/${projectName}/${locale}.${ext}`
      console.log(`Pulling s3://instructure-translations/${key} to ${filename}`)
      const { Body } = await s3.getObject({ Bucket, Key: key }).promise()
      mkdirp.sync(path.dirname(filename))
      writeFileSync(filename, Body.toString().replace(/^\uFEFF/, ''), 'utf8') // Strip BOM
    }
  }

  for (const project of projects) {
    if (program.project && project.name !== program.project) continue
    const folder = `imports/${project.name}`
    const files = await new Promise(resolve =>
      readdir(folder, (err, files) => resolve(files || []))
    )
    for (const file of files) {
      if (file.startsWith('.')) continue
      console.log(`Importing ${file} into ${project.location}`)
      if (project.location.endsWith('.xcodeproj')) {
        await run('xcodebuild', [
          '-importLocalizations',
          '-localizationPath',
          `translations/${folder}/${file}`,
          '-project',
          project.location
        ], { cwd: '..' })
      } else {
        await run('cp', [
          `${folder}/${file}`,
          `../${project.location}/${file}`
        ])
      }
    }
  }
}
